File: php/examples/04_webhook_receiver.php
<?php
/** Minimal receiver for the PHP built-in server: `php -S 0.0.0.0:8080 examples/04_webhook_receiver.php`
 *  Reads the RAW body (php://input) and verifies it before parsing. In Laravel use $request->getContent(). */
declare(strict_types=1);
require __DIR__ . '/_common.php';
use Amen\Config; use Amen\Webhooks\Handler;
$secret = Config::fromEnv(['apiKey' => getenv('AMN_API_KEY') ?: 'unused-for-receiver'])->webhookSecret ?? throw new RuntimeException('AMN_WEBHOOK_SECRET is required');
$handler = new Handler($secret, fn(array $e) => error_log("📩 {$e['type']} {$e['id']}: " . substr(json_encode($e['data']), 0, 300)));
[$status, $body] = $_SERVER['REQUEST_METHOD'] === 'POST' ? $handler->handle(getallheaders(), file_get_contents('php://input')) : [405, []];
http_response_code($status); header('Content-Type: application/json'); echo json_encode($body);

File: python/examples/04_webhook_receiver.py
#!/usr/bin/env python3
"""Minimal webhook receiver (stdlib only). Verifies the raw body, de-duplicates, logs the event.

    AMN_WEBHOOK_SECRET=... python examples/04_webhook_receiver.py   # listens on :8080/webhook
Expose it publicly (e.g. a tunnel) and register the URL with amen.webhooks.create(url).
"""
import json, os
from http.server import BaseHTTPRequestHandler, HTTPServer
from amen.config import Config
from amen.webhooks import WebhookHandler

Config.from_env  # loads .env
from dotenv import load_dotenv; load_dotenv(); load_dotenv("../.env")
secret = os.environ["AMN_WEBHOOK_SECRET"]
handler = WebhookHandler(secret, on_event=lambda e: print(f"📩 {e.type or '?'} {e.id}: {json.dumps(e.data)[:300]}"))


class H(BaseHTTPRequestHandler):
    def do_POST(self):
        raw = self.rfile.read(int(self.headers.get("Content-Length", 0)))
        code, body = handler.handle(dict(self.headers), raw)
        self.send_response(code); self.send_header("Content-Type", "application/json"); self.end_headers()
        self.wfile.write(json.dumps(body).encode())


print("listening on http://0.0.0.0:8080/webhook")
HTTPServer(("0.0.0.0", 8080), H).serve_forever()
